xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -187,6 +206,65 @@ void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
     adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
   }
 
+  @Override
+  void watchClusterData(String clusterName, ClusterWatcher watcher) {
+    checkNotNull(watcher, "watcher");
+    boolean needRequest = false;
+    if (!clusterWatchers.containsKey(clusterName)) {
+      logger.log(Level.FINE, "Start watching cluster {0}", clusterName);
+      needRequest = true;
+      clusterWatchers.put(clusterName, new HashSet<ClusterWatcher>());
+    }
+    Set<ClusterWatcher> watchers = clusterWatchers.get(clusterName);
+    if (watchers.contains(watcher)) {
+      logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
+    }
+    watchers.add(watcher);
+    // If local cache contains cluster information to be watched, notify the watcher immediately.
+    if (clusterNamesToClusterUpdates.containsKey(clusterName)) {
+      watcher.onClusterChanged(clusterNamesToClusterUpdates.get(clusterName));
+    }
+    if (rpcRetryTimer != null) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (needRequest) {
+      if (adsStream == null) {
+        startRpcStream();
+      }
+      adsStream.sendXdsRequest(ADS_TYPE_URL_CDS, clusterWatchers.keySet());
+    }
+  }
+
+  @Override
+  void cancelClusterDataWatch(String clusterName, ClusterWatcher watcher) {
+    checkNotNull(watcher, "watcher");
+    Set<ClusterWatcher> watchers = clusterWatchers.get(clusterName);
+    if (watchers == null || !watchers.contains(watcher)) {
+      logger.log(Level.WARNING, "Watcher {0} was not registered", watcher);
+      return;
+    }
+    watchers.remove(watcher);
```

Comment:
Is there a problem here? In `watchEndpointData` you are accepting multiple calls for the same `watcher` (with a warning) but without keeping a ref-count. However in this function you are removing the `watcher` on the first call.

We need to be very clear about the semantics here: if you are allowing multiple registrations of the same watcher you should keep a ref-count and require the same number of `cancel` calls to remove that watcher to avoid potential bugs.

But if the above is not desirable (which I agree) then you should not allow multiple registrations of the same watcher and the watchers then have to clean up their code to avoid this situation.

Reply:
Yes, thanks for pointing this out. I would make adding and removing consistent in semantics: adding the same watcher for the same cluster more than once is a no-op (with a warning log. Similarly, trying to remove a non-existing watcher is also no-op (with a warning).

Reply:
Changed. See above.

Reply:
> ... Similarly, trying to remove a non-existing watcher is also no-op (with a warning).

You are still leaving some scope for errors (in caller code not yours). Let's say the caller code has 2 adds and 2 removes for the same watcher (for whatever reason: 2 different people wrote the code etc). But you are removing the watcher on the first remove call leaving the user of the second watcher high and dry. Wouldn't it be better to generate an error when someone tries to add the same watcher twice? Like you said "why would anyone do that" in which case it's okay to do it

Reply:
As offline discussion, we will implement add/remove semantics same as Java's `Set` API.

Reply:
I updated method specification to state so.

Reply:
To make it same as `Set`, you should return true/false based on what you did.

Reply:
Hmm... I don't think that's necessary.

`Set` is a very very general thing, its API designers want to provide as much information as they could for different varieties of usage. The information of whether an addition or removal succeeded is useful for `Set` users to know if the element exists. But for us, it's not the case. We expect users for these APIs know what they are doing. That is, they are expected to know if a watcher has been added or removed, instead of relying on return value of add/remove API calls. We are aiming to provide APIs with strong specifications that do not put too much burden on callers (they don't need to worry about breaking things if cancelling more than once), but not to provide them such a usage to let them know if a watcher is added or removed.